src/query/sql/src/executor/physical_plans/physical_asof_join.rs

```
@@ -0,0 +1,416 @@
+use crate::executor::PhysicalPlan;
+use crate::executor::PhysicalPlanBuilder;
+use crate::optimizer::ColumnSet;
+use crate::optimizer::RelExpr;
+use crate::optimizer::SExpr;
+use crate::plans::BoundColumnRef;
+use crate::plans::ComparisonOp;
+use crate::plans::ConstantExpr;
+use crate::plans::FunctionCall;
+use crate::plans::Join;
+use crate::plans::JoinType;
+use crate::plans::LagLeadFunction;
+use crate::plans::ScalarExpr;
+use crate::plans::ScalarItem;
+use crate::plans::Window;
+use crate::plans::WindowFunc;
+use crate::plans::WindowFuncFrame;
+use crate::plans::WindowFuncFrameBound;
+use crate::plans::WindowFuncFrameUnits;
+use crate::plans::WindowFuncType;
+use crate::plans::WindowOrderBy;
+use crate::Visibility;
+
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+pub struct AsofJoin {
+    // A unique id of operator in a `PhysicalPlan` tree, only used for display.
+    pub plan_id: u32,
+    pub left: Box<PhysicalPlan>,
+    pub right: Box<PhysicalPlan>,
+    pub non_equi_conditions: Vec<RemoteExpr>,
+    // Now only support inner join, will support left/right join later
+    pub join_type: JoinType,
+
+    pub output_schema: DataSchemaRef,
+
+    // Only used for explain
+    pub stat_info: Option<PlanStatsInfo>,
+}
+
+impl AsofJoin {
+    pub fn output_schema(&self) -> Result<DataSchemaRef> {
+        Ok(self.output_schema.clone())
+    }
+}
+
+impl PhysicalPlanBuilder {
+    pub async fn build_asof_join(
+        &mut self,
+        join: &Join,
+        s_expr: &SExpr,
+        required: (ColumnSet, ColumnSet),
+        mut range_conditions: Vec<ScalarExpr>,
+        mut other_conditions: Vec<ScalarExpr>,
+    ) -> Result<PhysicalPlan> {
+        let right_prop = RelExpr::with_s_expr(s_expr.child(0)?).derive_relational_prop()?;
+        let left_prop = RelExpr::with_s_expr(s_expr.child(1)?).derive_relational_prop()?;
+
+        if range_conditions.is_empty() {
+            return Err(ErrorCode::Internal("Missing inequality condition!"));
```

It's would be better to return these errors during bind phase.